airbyte-integrations/connectors/source-google-sheets/google_sheets_source/google_sheets_source.py

```diff
@@ -0,0 +1,128 @@
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+from apiclient import errors
+from typing import Generator
+
+from airbyte_protocol import AirbyteConnectionStatus, Status
+from airbyte_protocol import AirbyteCatalog
+from airbyte_protocol import AirbyteLogger
+from airbyte_protocol import AirbyteMessage, Type
+from airbyte_protocol import Source
+
+from .helpers import Helpers
+from .models.generated.spreadsheet_values import *
+from .models.generated.spreadsheet import *
+
+ROW_BATCH_SIZE = 200
+
+
+class GoogleSheetsSource(Source):
+    """
+    Spreadsheets API Reference: https://developers.google.com/sheets/api/reference/rest/v4/spreadsheets
+    """
+
+    def __init__(self):
+        super().__init__()
+
+    def check(self, logger: AirbyteLogger, config_container) -> AirbyteConnectionStatus:
+        # Check involves verifying that the specified spreadsheet is reachable with our credentials.
+        config = config_container.rendered_config
+        client = Helpers.get_authenticated_sheets_client(config)
+        spreadsheet_id = config['spreadsheet_id']
+        try:
+            # Attempt to get first row of sheet
+            client.get(spreadsheetId=spreadsheet_id, includeGridData=False, ranges='1:1').execute()
+        except errors.HttpError as err:
+            reason = str(err)
+            # Give a clearer message if it's a common error like 404.
+            if err.resp.status == 404:
+                reason = "Requested spreadsheet was not found."
+
+            print(f"Formatted error: {reason}")
+            return AirbyteConnectionStatus(status=Status.FAILED, message=str(reason))
+
+        return AirbyteConnectionStatus(status=Status.SUCCEEDED)
+
+    def discover(self, logger: AirbyteLogger, config_container) -> AirbyteCatalog:
+        config = config_container.rendered_config
+        client = Helpers.get_authenticated_sheets_client(config)
+        spreadsheet_id = config['spreadsheet_id']
+        try:
+            logger.info(f"Running discovery on sheet {spreadsheet_id}")
+            spreadsheet_metadata = Spreadsheet.parse_obj(client.get(spreadsheetId=spreadsheet_id, includeGridData=False).execute())
+            sheet_names = [sheet.properties.title for sheet in spreadsheet_metadata.sheets]
+            streams = []
+            for sheet_name in sheet_names:
```

Comment:
to make sure i understand this is for a single google sheet the individual pages inside of it?

Reply:
correct. The terminology is Spreadsheet for the "entire sheet", and subsheets are called a Sheet